balances-tx-payment/rpc/src/lib.rs

```diff
@@ -0,0 +1,186 @@
+{
+	#[method(name = "encointer_queryAssetFeeDetails", blocking)]
+	fn query_asset_fee_details(
+		&self,
+		asset_id: AssetId,
+		encoded_xt: Bytes,
+		at: Option<BlockHash>,
+	) -> RpcResult<FeeDetails<NumberOrHex>>;
+}
+
+pub struct BalancesTxPaymentRpc<C, Q, R, S> {
+	client: Arc<C>,
+	_marker: std::marker::PhantomData<(Q, R, S)>,
+}
+
+impl<C, Q, R, S> BalancesTxPaymentRpc<C, Q, R, S> {
+	pub fn new(client: Arc<C>) -> Self {
+		BalancesTxPaymentRpc { client, _marker: Default::default() }
+	}
+}
+
+impl<C, Block, AssetId, Balance, AssetBalance>
+	BalancesTxPaymentApiServer<<Block as BlockT>::Hash, AssetId>
+	for BalancesTxPaymentRpc<C, Block, Balance, AssetBalance>
+where
+	AssetId: 'static + Clone + Copy + Encode + Decode + Send + Sync + PartialEq,
+	AssetBalance: 'static
+		+ Clone
+		+ Encode
+		+ Decode
+		+ Send
+		+ Sync
+		+ PartialEq
+		+ Into<NumberOrHex>
+		+ MaybeDisplay
+		+ Copy,
+	Block: BlockT,
+	C: ProvideRuntimeApi<Block> + HeaderBackend<Block> + Send + Sync + 'static,
+	C::Api: BalancesTxPaymentApiRuntimeApi<Block, Balance, AssetId, AssetBalance>,
+	C::Api: TransactionPaymentRuntimeApi<Block, Balance>,
+	Balance:
+		Codec + MaybeDisplay + Copy + TryInto<NumberOrHex> + Send + Sync + 'static + From<u128>,
+{
+	fn query_asset_fee_details(
+		&self,
+		asset_id: AssetId,
+		encoded_xt: Bytes,
+		at: Option<Block::Hash>,
+	) -> RpcResult<FeeDetails<NumberOrHex>> {
+		let api = self.client.runtime_api();
+		let at = BlockId::hash(at.unwrap_or_else(|| self.client.info().best_hash));
+
+		let encoded_len = encoded_xt.len() as u32;
+
+		let uxt: Block::Extrinsic = Decode::decode(&mut &*encoded_xt).map_err(|e| {
+			CallError::Custom(ErrorObject::owned(
+				Error::DecodeError.into(),
+				"Unable to query fee details.",
+				Some(format!("{:?}", e)),
+			))
+		})?;
+		let fee_details = api.query_fee_details(&at, uxt, encoded_len).map_err(|e| {
+			CallError::Custom(ErrorObject::owned(
+				Error::RuntimeError.into(),
+				"Unable to query fee details.",
+				Some(e.to_string()),
+			))
+		})?;
+
+		let try_into_rpc_balance = |value: AssetBalance| {
+			value.try_into().map_err(|_| {
+				JsonRpseeError::Call(CallError::Custom(ErrorObject::owned(
+					ErrorCode::InvalidParams.code(),
+					format!("{} doesn't fit in NumberOrHex representation", value),
+					None::<()>,
+				)))
+			})
+		};
+
+		Ok(FeeDetails {
+			inclusion_fee: if let Some(inclusion_fee) = fee_details.inclusion_fee {
+				let base_fee = api
+					.balance_to_asset_balance(&at, inclusion_fee.base_fee, asset_id)
+					.map_err(|e| {
+						CallError::Custom(ErrorObject::owned(
+							Error::RuntimeError.into(),
+							"Unable to query balance conversion.",
+							Some(e.to_string()),
+						))
+					})?
+					.map_err(|_e| {
+						CallError::Custom(ErrorObject::owned(
+							Error::RuntimeError.into(),
+							"Unable to query balance conversion.",
+							Some("Unable to query balance conversion."),
+						))
+					})?;
```

Comment:
I thought I had written this in the last review, but I seem to have missed it. I would love to see something like this here.

```rust
api.
    .balance_to_asset_balance(&at, inclusion_fee.len_fee, asset_id)
	.map_err(|e| runtime_error_into_rpc_error(e, "Unable to base fee")?;
```

We could re-use the function three times already. :)